File: web/src/text/isWithinCharacterLimit.ts
interface CharacterLimitOptions {
  limit: number
  value: string
}

const segmenter = new Intl.Segmenter('th', { granularity: 'grapheme' })

/** Checks a limit using visible Unicode characters, including Thai clusters. */
export default function isWithinCharacterLimit({ limit, value }: CharacterLimitOptions): boolean {
  let count = 0
  for (const _segment of segmenter.segment(value)) {
    count += 1
    if (count > limit) return false
  }
  return true
}

File: app/lib/text/is_within_character_limit.dart
import 'package:characters/characters.dart';

/// Reports whether [value] fits within [limit] visible Unicode characters.
///
/// Thai tone marks and vowels that combine with a base letter count as part of
/// the same character instead of consuming separate slots.
bool isWithinCharacterLimit({required String value, required int limit}) =>
    value.characters.length <= limit;
